src/java-runtime/utils/upstreamApi.ts

```diff
@@ -36,36 +39,117 @@ export async function resolveRequirements(): Promise<any> {
                 }
                 invalidJavaHome(reject, msg);
             }
-            javaVersion = await getJavaVersion(javaHome);
+            javaVersion = await getMajorVersion(javaHome);
+            toolingJre = javaHome;
+            toolingJreVersion = javaVersion;
         } else {
-            // java.home not specified, search valid JDKs from env.JAVA_HOME, env.PATH, Registry(Window), Common directories
-            const javaRuntimes = await findJavaHomes();
-            const validJdks = javaRuntimes.filter(r => r.version >= REQUIRED_JDK_VERSION);
-            if (validJdks.length > 0) {
-                sortJdksBySource(validJdks);
-                javaHome = validJdks[0].home;
-                javaVersion = validJdks[0].version;
+            // java.home not specified, search valid JDKs from env.JAVA_HOME, env.PATH, SDKMAN, jEnv, jabba, Common directories
+            const javaRuntimes = await findRuntimes({checkJavac: true, withVersion: true, withTags: true});
+            if (!toolingJre) { // universal version
+                // as latest version as possible.
+                sortJdksByVersion(javaRuntimes);
+                const validJdks = javaRuntimes.filter(r => r.version && r.version.major >= REQUIRED_JDK_VERSION);
+                if (validJdks.length > 0) {
+                    sortJdksBySource(validJdks);
+                    javaHome = validJdks[0].homedir;
+                    javaVersion = validJdks[0].version?.major ?? 0;
+                    toolingJre = javaHome;
+                    toolingJreVersion = javaVersion;
+                }
+            } else { // pick a default project JDK/JRE
+                /**
+                 * For legacy users, we implicitly following the order below to
+                 * set a default project JDK during initialization:
+                 * java.home > env.JDK_HOME > env.JAVA_HOME > env.PATH
+                 *
+                 * We'll keep it for compatibility.
+                 */
+                if (javaHome && (await getRuntime(javaHome) !== undefined)) {
```

Comment:
Can we directly call `await getRuntime(javaHome, {withVersion: true})` only once here and store the return value?

Reply:
yes we can, but `getRuntime` is cheap. Here I simply copy&paste to use the same logic in redhat.java.

Reply:
We can optimize it on both here and upstream later, if necessary.
e.g.

```ts
const runtime = javaHome ? await getRuntime(javaHome, { withVersion: true }) : undefined;
if (runtime) {
    javaHome = runtime.homedir;
    javaVersion = runtime.version?.major ?? 0;
} else if (javaRuntimes.length) {
...
```

Reply:
I'm OK if it's cheap. never mind.